assets/src/edit-story/app/font/fontProvider.js

```diff
@@ -23,23 +23,72 @@ import { __ } from '@wordpress/i18n';
  * External dependencies
  */
 import PropTypes from 'prop-types';
-import { useCallback, useState, useRef } from 'react';
+import { useCallback, useState, useRef, useEffect } from 'react';
 
 /**
  * Internal dependencies
  */
 import loadStylesheet from '../../utils/loadStylesheet';
+import { useStory } from '../story';
 import Context from './context';
 import useLoadFonts from './effects/useLoadFonts';
 import useLoadFontFiles from './actions/useLoadFontFiles';
+import useGetRecentFonts from './useGetRecentFonts';
 
 const GOOGLE_MENU_FONT_URL = 'https://fonts.googleapis.com/css';
 
 function FontProvider({ children }) {
+  const { currentPage, pages } = useStory((state) => ({
+    currentPage: state.state.currentPage,
+    pages: state.state.pages,
+  }));
   const [fonts, setFonts] = useState([]);
 
+  const getRecentFonts = useGetRecentFonts();
+  const [recentFonts, setRecentFonts] = useState([]);
+
+  const countRef = useRef({
+    elementCount: currentPage?.elements?.length
+      ? currentPage.elements.filter(({ type }) => type === 'text').length
+      : 0,
+    pageCount: pages?.length || 1,
+  });
+
+  /*
+   * This effect checks if the number of text elements on the current page has changed
+   * or if a page has been added/removed and updates the recent fonts if yes.
```

Comment:
@barklund WDYT of this approach instead of trying to track the font count in different places in the code (as the issue implementation brief suggests, e.g. when removing and adding a text element / changing font / changing multiple text elements at a time). This seems a bit more manageable compared to making sure to track in various places in the code, however, it will create some extra counting (e.g. when removing a Page which didn't have any text element / adding an empty page).

Thoughts?

Reply:
What happens if I change a font on an existing text field then? How would this run?

Also, if you loop through all the elements (`currentPage.elements.filter()`) on every change to `pages` or `currentPage`, you might as well just count all the fonts anyway - there's almost no difference?

Reply:
> What happens if I change a font on an existing text field then? How would this run?

`updateRecentFonts` is used in the `fontPicker` currently directly.

> Also, if you loop through all the elements (currentPage.elements.filter()) on every change to pages or currentPage, you might as well just count all the fonts anyway - there's almost no difference?

Well, it only counts all the fonts if
a) The number of the _**text**_ elements changes on the current page, or
b) The number of the total pages changes

Meaning that it doesn't run on _every_ change. E.g. changing an image or adding a new image/shape, etc. doesn't run it. Thoughts?

Reply:
@barklund FYI: Will rework this PR since apparently the information in https://github.com/google/web-stories-wp/issues/1955 was not correct.

Talked to Sam and we need to display only the recently used fonts from that page reload -- this is not based on the previously used fonts in the story at all. Will update the description of the issue, too.